### PR TITLE
PhraserBeam

### PR DESCRIPTION
**Pull request by jamesgorrie**
The explanation of the new query can be [found in the overview of the test](https://github.com/wellcomecollection/catalogue/pull/459/files#diff-385960895593ee2dc4585834f1f0c04fR3).

I have also split the queries up into
- `PartialQuery`
- `ComboQuery`

This is to help with how we structure the queries, but to ensure we aren't surfacing those to the client.

The tests are also now structured into the basic query tests, matching on title, genres, IDs etc, and runs through and queries we are running tests on to make sure we don't break that. 

The scoring tests however are more to explain what the queries _should_ be doing, but it's not possible without the real data set as scoring relies so heavily on it.

We will be running session with the different collections team to help us start creating more qualitative testing, probably using the new search bar:
Screenshot 2020-03-06 at 09 11 09

### CODE REVIEW

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/ElasticsearchQuery.scala [hunk 1]
```diff
@@ -2,35 +2,104 @@ package uk.ac.wellcome.platform.api.elasticsearch
...
   val q: String
   val elasticQuery: Query
 }
 
-// Queries sent to the application _should_ implement this class
-// as it wraps core functionality that we wouldn't want to break.
-// We haven't enforced this in the type system as we _might_ want
-// to try more extreme tests.
+trait ElasticsearchPartialQuery extends ElasticsearchQuery
+trait ElasticsearchComboQuery extends ElasticsearchQuery
```
Comment (jamieparkinson) [inline api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/ElasticsearchQuery.scala]: Are these traits actually used anywhere? It looks like they're just used for tagging different query types, but I can't find anywhere where those tags are used

File: docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Test 6 - `ConstScore` vs `BoolBoosted`
+
+## Overview
+
+Currently we are matching _named features_<sup>*</sup> with a simple **AND** query.
```
Comment (warrd) [inline docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md]: Might be nice to link to bellow like this, took me a second to find what this refers to: https://stackoverflow.com/a/48453745

File: api/api/src/test/scala/uk/ac/wellcome/platform/api/elasticsearch/QueryVariantsTest.scala [hunk 3]
```diff
@@ -143,6 +163,8 @@ class QueryVariantsTest
         fieldSort("canonicalId").order(SortOrder.ASC))
       .explain(true)
 
+    println(request.show)
```
Comment (warrd) [inline api/api/src/test/scala/uk/ac/wellcome/platform/api/elasticsearch/QueryVariantsTest.scala]: Forgot about this?

File: docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md (hunk 2 above)
Comment (jamesgorrie) [reply docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md]: https://github.com/wellcomecollection/catalogue/pull/459/commits/9c25387522a1255826ac09dd106f100e2c6cf9de

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/ElasticsearchQuery.scala (hunk 1 above)
Comment (jamesgorrie) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/ElasticsearchQuery.scala]: Ah yes - this is exactly what I was trying to enforce via the type system and forgot.
https://github.com/wellcomecollection/catalogue/pull/459/commits/b54074ddc4abe2ddd603723103fb9dde8a4b835e

Tests didn't pick it up as the previous queries had the core query in, and they did the core amount of work, which is great.

File: docs/SUMMARY.md [hunk 4]
```diff
@@ -11,6 +11,7 @@
     - [Test 5 - Scoring Tiers](search_relevance/tests/005_stacked_queries.md)
     - [Test 6 - English tokeniser and Contributors](search_relevance/tests/006-english-tokeniser-and-contributors.md)
     - [Test 7 - BoolBoosted vs ConstScore](search_relevance/tests/007-ConstScore-vs-BoolBoosted.md)
+    - [Test 7 - BoolBoosted vs PhaserBeam](search_relevance/tests/007-BoolBoosted-vs-PhaserBeam.md)
```
Comment (kenoir) [inline docs/SUMMARY.md]: 👏

File: docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# Test 6 - `ConstScore` vs `BoolBoosted`
```
Comment (kenoir) [inline docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md]: Name?

File: docs/SUMMARY.md (hunk 4 above)
Comment (kenoir) [reply docs/SUMMARY.md]: Should this be `008-`?

Comment (jamesgorrie) [reply docs/SUMMARY.md]:
```suggestion
    - [Test 8 - BoolBoosted vs PhaserBeam](search_relevance/tests/007-BoolBoosted-vs-PhaserBeam.md)
```

File: docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md (hunk 5 above)
Comment (jamesgorrie) [reply docs/search_relevance/tests/008-BoolBoosted-vs-PhaserBeam.md]:
```suggestion
# Test 008 - BoolBoosted vs PhaserBeam
```

**Pull request merged**